### PR TITLE
[ REVIEW ] Adding serving, BYOData, ML workflow refactor

### PR DESCRIPTION
**Pull request by miroenev**
Added support for on-demand inference 

> via Flask server + ForestInference for prediction, see aws/code/serve.py

BYOData support and documentation 

> feature and label enumeration in HPODatasets.py, +CSV support, additional documentation in notebook

ML Workflow refactor 
> each code path is its own class/file in aws/code/workflows/*.py 

Separated out SageMaker specific configuration parsing 
> see aws/code/HPOConfig.py

Job-name string parsing replaced by reading ENV variables set in Dockerfile

Updated container references to 0.15 using base vs runtime, and added 0.16 nightly

### CODE REVIEW

File: aws/code/workflows/MLWorkflowMultiCPU.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        print('> score predictions')
+        # y_test = y_test.compute()
+        score = accuracy_score ( y_test.astype( self.hpo_config.dataset_dtype ),
+                                predictions.astype( self.hpo_config.dataset_dtype ) )
+
+        print(f'\t score = {score}')
+        self.cv_fold_scores += [score]
+        return score
+
+    def save_best_model ( self, score, trained_model, filename = 'saved_model' ):
```
Comment (wphicks) [inline aws/code/workflows/MLWorkflowMultiCPU.py:181]: Similar to earlier comment: Do we want to determine best model on the basis of a single fold, or should we evaluate across all k?

File: aws/code/workflows/MLWorkflowMultiGPU.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    
+    @timer_decorator
+    def cluster_initialize ( self ):
+        """ Initialize the dask compute cluster based on the number of available GPU workers.
+            XGBoost has a known issue where training fails if any worker has no data partition
+            so when initializing a dask cluster for xgboost we may need to limit the number of workers
+            see 3rd limitations bullet @ https://xgboost.readthedocs.io/en/latest/tutorials/dask.html 
+        """
+
+        cluster = None;  client = None
```
Comment (wphicks) [inline aws/code/workflows/MLWorkflowMultiGPU.py:59]: Let's move these to separate lines. See the note on compound statements in https://www.python.org/dev/peps/pep-0008/#other-recommendations.

I noticed a few other compound statements earlier but didn't call them out because I'd forgotten that PEP8 officially discourages them. We might clean those up as well.

File: aws/code/workflows/MLWorkflowMultiGPU.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        X_train = train.drop(label_column, axis = 1)
+        y_test = test[label_column]
+        X_test = test.drop(label_column, axis = 1)
+
+        # force execution
+        X_train, y_train, X_test, y_test = persist_across_workers( self.client,
+                                                                    [ X_train, y_train, X_test, y_test ], 
+                                                                    workers = self.client.has_what().keys() )        
+
+        return X_train.astype( self.hpo_config.dataset_dtype ),\
```
Comment (wphicks) [inline aws/code/workflows/MLWorkflowMultiGPU.py:132]: Trailing backslashes; see above. I won't specifically flag remaining examples.

File: aws/code/workflows/MLWorkflowMultiGPU.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    @timer_decorator
+    def score ( self, y_test, predictions ): 
+        """ Score predictions vs ground truth labels on test data """
+        print('> score predictions')
+        y_test = y_test.compute()
+        score = accuracy_score ( y_test.astype( self.hpo_config.dataset_dtype ),
+                                predictions.astype( self.hpo_config.dataset_dtype ) )
+
+        print(f'\t score = {score}')
+        self.cv_fold_scores += [score]
```
Comment (wphicks) [inline aws/code/workflows/MLWorkflowMultiGPU.py:180]: Prefer `append` to `+` operator for building up a list.

File: aws/code/workflows/MLWorkflowMultiGPU.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        return score
+
+    def save_best_model ( self, score, trained_model, filename = 'saved_model' ): 
+        """  Persist/save model that sets a new high score """
+
+        if score > self.best_score:
+            self.best_score = score
+            print('> saving high-scoring model')
+            output_filename = self.hpo_config.model_store_directory + '/' + str( filename )
```
Comment (wphicks) [inline aws/code/workflows/MLWorkflowMultiGPU.py:189]:
```suggestion
            output_filename = os.path.join(self.hpo_config.model_store_directory, filename)
```

File: aws/code/workflows/MLWorkflowMultiGPU.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            if 'XGBoost' in self.hpo_config.model_type:
+                trained_model.save_model( output_filename + '_mgpu_xgb' )
```
Comment (wphicks) [inline aws/code/workflows/MLWorkflowMultiGPU.py:191]: Prefer `.format`, f-strings, or `join` to `+` operator for concatenating strings

File: aws/helper_functions.py [hunk 7]
```diff
@@ -116,19 +71,76 @@ def download_best_model( bucket, s3_model_output, hpo_results, local_directory )
     """ Download best model from S3"""
     try:
         target_bucket = boto3.resource('s3').Bucket( bucket )
-        path_prefix = s3_model_output.split('/')[3] + '/' + hpo_results['BestTrainingJob']['TrainingJobName'] + '/output/'
+        path_prefix = s3_model_output.split('/')[-1] + '/' + hpo_results['BestTrainingJob']['TrainingJobName'] + '/output/'
```
Comment (wphicks) [inline aws/helper_functions.py:74]: For all of the path manipulation that's going on in this function, I'd recommend using [pathlib](https://docs.python.org/3/library/pathlib.html), but we probably want to at least avoid `+` for string concatenation.

File: aws/helper_functions.py [hunk 8]
```diff
@@ -116,19 +71,76 @@ def download_best_model( bucket, s3_model_output, hpo_results, local_directory )
...
+            validate_region ( region )
+        elif dataset_directory in [ 'NYC_taxi']:
+            data_choice_str = 'nyc'
+            validate_region ( region )
+        else:
+            data_choice_str = 'byo'
+            
+        code_choice_str = code_choice[0] + code_choice[-3:]
+        
+        if 'randomforest' in algorithm_choice.lower() : algorithm_choice_str = 'RF'
```
Comment (wphicks) [inline aws/helper_functions.py:111]: Similarly to compound statements with semicolons, PEP8 discourages this.

File: aws/helper_functions.py [hunk 9]
```diff
@@ -116,19 +71,76 @@ def download_best_model( bucket, s3_model_output, hpo_results, local_directory )
...
+        if 'xgboost' in algorithm_choice.lower() : algorithm_choice_str = 'XGB'    
+        
+        instance_type_str = '-'.join( instance_type.split('.')[1:] )        
+
+        random_8char_str = ''.join( random.choices( uuid.uuid4().hex, k=8 ) )
```
Comment (wphicks) [inline aws/helper_functions.py:116]: Are we just looking to construct a unique job name? If so, I'd keep the full uuid. If not, is there any reason to think that randomly choosing 8 characters from the uuid is better than merely truncating it to 8 characters?

File: aws/helper_functions.py [hunk 10]
```diff
@@ -116,19 +71,76 @@ def download_best_model( bucket, s3_model_output, hpo_results, local_directory )
...
+        job_name = f"{data_choice_str}-{code_choice_str}" \
+                    f"-{algorithm_choice_str}-{cv_folds}cv"\
+                    f"-{instance_type_str}-{random_8char_str}"
+        
+        job_name = job_name[:trim_limit]
+        
+        print ( f'generated job name : {job_name}\n')
+        
+    except Exception as error: 
+        print( f'ERROR: unable to generate job name: {error}' )
```
Comment (wphicks) [inline aws/helper_functions.py:127]: Use `traceback.print_exc` or `traceback.format_exc` here.

File: aws/helper_functions.py [hunk 11]
```diff
@@ -116,19 +71,76 @@ def download_best_model( bucket, s3_model_output, hpo_results, local_directory )
...
+        
+    if region not in ['us-east-1', 'us-west-2']:
+        raise Exception ( 'Unsupported region based on demo data location,'\
+                          ' please switch to us-east-1 or us-west-2' )
+        
+def get_notebook_path():
+    """ Query the root notebook directory [ once ], 
+        we'll end up placing our best trained model in this location.
+    """
+    try: working_directory
```
Comment (wphicks) [inline aws/helper_functions.py:144]: Sorry, I'm not sure what this code is supposed to do. It looks to me like it will always call and return `os.getcwd()`.

File: aws/code/MLWorkflow.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+import functools
+import time
+
+class MLWorkflow ( ):
+    def __init__ ( self ):
+        return None
+
+    def create_workflow( self, hpo_config ):        
+        if hpo_config.compute_type == 'single-CPU':
+            from workflows.MLWorkflowSingleCPU import MLWorkflowSingleCPU
```
Comment (wphicks) [inline aws/code/MLWorkflow.py:11]: Any particular reason for the delayed import here?

File: aws/code/entrypoint.sh [hunk 13]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+source activate rapids
+if [[ "$1" == "serve" ]]; then
+    echo -e "@ entrypoint -> launching serving script \n"
+    python serve.py
```
Comment (wphicks) [inline aws/code/entrypoint.sh:5]: I know we talked about this a bit previously, but it's probably worth revisiting whether or not this should be deployed through a WSGI server for the use case(s) we intend.

**Review by wphicks — CHANGES_REQUESTED** [review]
This seems like a nice refactor, @miroenev! I've made a bunch of in-line comments, but there are also a few higher level things I wanted to address.

1. Like I mentioned to you offline, there are a bunch of linting errors throughout. I started to comment on these, but there were enough that it probably makes sense for you to just run it through standard linters yourself. [hadolint](https://github.com/hadolint/hadolint) is great for Dockerfiles, and any of the usual suspects will work for Python.
2. We're doing a whole bunch of logging with print statements. Might we instead use an actual logging system so that we can (among other things) more easily control the logging level?
3. There are a bunch of print statements that end with a newline (and sometimes start with a newline as well). Is there any particular reason other than adding whitespace to the logging output?
4. There are quite a few places with unnecessary (and potentially counterproductive) calls to `str`. I flagged a few of them, but it might be worth grepping for `str(` and make sure that you really need it where it's being used.

File: aws/code/HPOConfig.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+                compute_type = 'multi-CPU'
+            elif 'singlecpu' in compute_selection:
+                compute_type = 'single-CPU'
+            elif 'singlegpu' in compute_selection:
+                compute_type = 'single-GPU'                
+                
+            # parse CV folds
+            cv_folds = int( os.environ['CV_FOLDS'] )
+
+        except Exception as error:
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/HPOConfig.py:83]: Good question, I don't think we should see partial configurations so a single try-catch should be sufficient but I'm curious to discuss with the team. And agree that we should increase the specificity of the Exception to KeyError.

File: aws/code/MLWorkflow.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    def emit_final_score ( self ): pass
+
+
+def timer_decorator ( target_function ):
+
+    @functools.wraps ( target_function )
+    def timed_execution_wrapper ( *args, **kwargs ):
+        start_time = time.perf_counter()
+        result = target_function ( *args, **kwargs )
+        print(f" --- {target_function.__name__} completed in {round( time.perf_counter() - start_time, 5 )} s")
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/MLWorkflow.py:60]: I agree that we should split up the timing from the print and use the format string instead of round.

File: aws/code/serve.py [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+        Load [ XGBoost or RandomForest ] model into memory 
+        This function scans the model SageMaker model directory and parses saved model names
+        """
+        xgb_models = glob.glob('/opt/ml/model/*_xgb')
+        rf_models = glob.glob('/opt/ml/model/*_rf')
+        print( f'detected xgboost models : {xgb_models}' ) 
+        print( f'detected randomforest models : {rf_models}\n\n' ) 
+        model_type = None
+
+        start_time = time.time()
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/serve.py:69]: he functions I've measured with time.time are generally those that are not very compute focused (e.g., writing to disk) or when capturing a large body of smaller functions that already have perf_counters internally. For the purpose of consistency, I think its worth picking one timer type (likely perf_counter) and sticking with it.

File: aws/helper_functions.py (hunk 11 above)
Comment (miroenev) [reply aws/helper_functions.py:144]: This is a remnant from the previous notebook, the idea being that you only check the working directory once (if it already exists skip the check).

File: aws/code/MLWorkflow.py (hunk 12 above)
Comment (miroenev) [reply aws/code/MLWorkflow.py:11]: Without the delay, it becomes a circular import (i.e. in the child class there is an import of the parent which is not fully defined yet). Maybe there is a better way to do this?

File: aws/code/entrypoint.sh (hunk 13 above)
Comment (miroenev) [reply aws/code/entrypoint.sh:5]: Agreed, seems like we should schedule a deeper dive conversation around the server (including front-loading the FIL model, etc).

File: aws/code/train.py [hunk 17]
```diff
@@ -14,59 +14,55 @@
...
+        # handle missing samples [ drop ]
+        dataset = ml_workflow.handle_missing_data ( dataset )
 
-            # run ETL [  ingest -> repartition -> drop missing -> split -> persist ]
-            X_train, X_test, y_train, y_test = rapids_sagemaker.ETL ( columns = dataset_feature_columns, 
-                                                                      label_column = dataset_label_column,
-                                                                      random_seed = i_fold ) 
+        # split into train and test set
+        X_train, X_test, \
+            y_train, y_test = ml_workflow.split_dataset ( dataset, random_state = i_fold )
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/train.py:37]: This is a good point that I think would make sense bringing up to the team. Perhaps an augmentation to cuml?

File: aws/code/train.py [hunk 18]
```diff
@@ -14,59 +14,55 @@
...
-        rapids_sagemaker.emit_final_score()
-                        
-        print( f'total elapsed time = { round( time.time() - start_time) } seconds\n' )
-    
-        sys.exit(0) # success exit code
+        # save trained model [ if it sets a new-high score ]
+        ml_workflow.save_best_model( score, trained_model )
 
-    except Exception as error:
+        # restart cluster to avoid memory creep [ for multi-CPU/GPU ]
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/train.py:51]: The dask cluster memory used to creep up after repeated runs in the cross-validation loop. I think there may have been a small bug somewhere in the RF code. The cluster reset was a way of preventing the creep for larger cross-validation selections, its quite fast though maybe worth disabling as a testing step before new releases.

File: aws/code/train.py [hunk 19]
```diff
@@ -14,59 +14,55 @@
...
+
+        sys.exit(0) # success exit code
+
+    except Exception as error:        
         trc = traceback.format_exc()           
-        print( ' ! exception: ' + str(error) + '\n' + trc, file = sys.stderr)
-        
-        sys.exit(-1) # a non-zero exit code causes the training job to be marked as failed
\ No newline at end of file
+        print( f' ! exception: {str(error)} \n {trc}')
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/train.py:66]: I'll try that, I recall having a reason for the current implementation but will double check.

File: aws/code/workflows/MLWorkflowMultiCPU.py [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+        if score > self.best_score:
+            self.best_score = score
+            print('> saving high-scoring model')
+            output_filename = self.hpo_config.model_store_directory + '/' + str( filename )
+            if 'XGBoost' in self.hpo_config.model_type:
+                trained_model.save_model( output_filename + '_mcpu_xgb' )
+            elif 'RandomForest' in self.hpo_config.model_type:
+                joblib.dump ( trained_model, output_filename + '_mcpu_rf' )
+            
+    @timer_decorator
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/workflows/MLWorkflowMultiCPU.py:193]: Likely not, this is a good catch. Perhaps adding an async_timer_decorator that just says 'this is a lazy/async function so timing will be rolled into upstream computation'  or something more clear :]

File: aws/code/workflows/MLWorkflowMultiCPU.py (hunk 1 above)
Comment (miroenev) [reply aws/code/workflows/MLWorkflowMultiCPU.py:181]: Yeah its an interesting question, the tricky thing is do you keep all the models in memory or do you write them all to disk? I decided to go with only saving once a new high score is reached within the cross-validation folds, though I agree its a subjective choice.

File: aws/helper_functions.py (hunk 9 above)
Comment (miroenev) [reply aws/helper_functions.py:116]: I think this was to stay within the 32 character limit of SageMaker job names.

File: aws/code/train.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (wphicks) [reply aws/code/train.py:37]: Yeah, sklearn has `sklearn.cross_validation.KFold` which is a nice helper function for this sort of thing. If we can get a version that guarantees no GPU-host memory copies, that would be very nice.

File: aws/code/train.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (wphicks) [reply aws/code/train.py:51]: Great, thanks for the info! Feel free to mark this resolved.

File: aws/code/train.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (wphicks) [reply aws/code/train.py:66]: Yeah, no worries if `print_exc` doesn't do what you want. I'd at least pull off the `str()` call around the exception object if you don't go with `print_exc`. You don't need it in this f-string.

File: aws/code/workflows/MLWorkflowMultiCPU.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (wphicks) [reply aws/code/workflows/MLWorkflowMultiCPU.py:193]: Cool, sounds good! I wouldn't even worry about having a separate async decorator unless you think that's going to be a point of confusion. You can probably just leave the function undecorated.

File: aws/code/workflows/MLWorkflowMultiCPU.py (hunk 1 above)
Comment (wphicks) [reply aws/code/workflows/MLWorkflowMultiCPU.py:181]: Sorry, looking back over this, I was actually misinterpreting what we're doing with the cross-validation here. Feel free to mark this conversation as resolved, but I may touch base with you offline to make sure I'm correctly understanding the intent. That's for my own education, though; this bit looks fine.

File: aws/helper_functions.py (hunk 9 above)
Comment (wphicks) [reply aws/helper_functions.py:116]: Sorry, let me rephrase with better context: it looks like we're already trimming down to 32 characters at line 122. Is there any reason to "pre-trim" this portion of the name?

Also, same question about the random selection of characters from the generated uuid.

File: aws/code/MLWorkflow.py (hunk 12 above)
Comment (wphicks) [reply aws/code/MLWorkflow.py:11]: Ah, of course! In that case, I'd pull this out as a separate factory function to break the circular dependency. I just noticed that my linter recommends that anyway, since this method makes no use `self` or `cls`.

File: aws/helper_functions.py (hunk 11 above)
Comment (wphicks) [reply aws/helper_functions.py:144]: Ah, I thought that might be what it was trying to do. Unfortunately, this does not actually cache the result. Here's a little snippet to demonstrate:

```python
def incorrect_cache():
    try:
        return result
    except NameError:
        print('Cache miss')
        result = 2 + 2
        return result

def correct_cache():
    try:
        return correct_cache.result
    except AttributeError:
        print('Cache miss')
        correct_cache.result = 2 + 2
        return correct_cache.result


if __name__ == '__main__':
    print('Incorrect cache demonstration:')
    for i in range(10):
        print("{}: {}".format(i, incorrect_cache()))
    print('Correct cache demonstration:')
    for i in range(10):
        print("{}: {}".format(i, correct_cache()))
```

And the output:
```bash
Incorrect cache demonstration:
Cache miss
0: 4
Cache miss
1: 4
Cache miss
2: 4
Cache miss
3: 4
Cache miss
4: 4
Cache miss
5: 4
Cache miss
6: 4
Cache miss
7: 4
Cache miss
8: 4
Cache miss
9: 4
Correct cache demonstration:
Cache miss
0: 4
1: 4
2: 4
3: 4
4: 4
5: 4
6: 4
7: 4
8: 4
9: 4
```
The scope of `working_directory` is local to the function, so referencing it at line 144 is *always* a `NameError`.

Comment (wphicks) [reply aws/helper_functions.py:144]: And I should probably add that the best way to cache is with one of `functools` [caching decorators](https://docs.python.org/3/library/functools.html#functools.cache) unless you need something with more logic to it.

File: aws/rapids_sagemaker_hpo.ipynb [hunk 21]
```diff
@@ -1 +1,1359 @@
...
+   "metadata": {},
+   "source": [
+    "<img src='img/results.png' width='70%'>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In the case of the CPU runs, 12 jobs were stopped since they exceeded the 24 hour limit we set. <a href='img/cpu_hpo_100x10.png'>CPU Job Summary Image</a>\n",
```
Comment (zronaghi) [inline aws/rapids_sagemaker_hpo.ipynb:1274]: Total CPU runtime is more than 3 days? Would be good to add a mention or reference to this, at the beginning of the notebook as well

**Review by zronaghi — COMMENTED** [review]
Overall nice refactor and easier to modify for additional datasets/algorithms. Great feedback by Will, just a few additional comments:

File: aws/code/workflows/MLWorkflowMultiCPU.py [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+        self.dataset_cache = None
+        
+        self.cv_fold_scores = []
+        self.best_score = -1
+
+        self.cluster, self.client = self.cluster_initialize()
+    
+    @timer_decorator
+    def cluster_initialize ( self ):
+        """ Initialize the dask compute cluster based on the number of available GPU workers.
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/workflows/MLWorkflowMultiCPU.py:52]: Good catch!

File: aws/code/workflows/MLWorkflowMultiCPU.py [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+        """ Emit score for parsing by the cloud HPO orchestrator """
+
+        print(f'total_time = {round( time.time() - self.start_time )} seconds ')
+
+        if self.hpo_config.cv_folds > 1 :
+            print(f'fold scores : {self.cv_fold_scores} \n')
+
+        final_score = sum(self.cv_fold_scores) / len(self.cv_fold_scores) # average
+
+        print(f'final-score: {final_score}; \n')
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/workflows/MLWorkflowMultiCPU.py:216]: That might be useful, though kind of against the principle of cross-validation. Let me know if you'd like it and I'd be happy to log/expose the best score.

File: aws/code/workflows/MLWorkflowSingleCPU.py [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+            trained_model = xgboost.train( dtrain = dtrain, params = self.hpo_config.model_params,
+                                           num_boost_round = self.hpo_config.model_params['num_boost_round'] )
+
+        elif 'RandomForest' in self.hpo_config.model_type:
+            print('> fit randomforest model')
+            trained_model = RandomForestClassifier ( n_estimators = self.hpo_config.model_params['n_estimators'],
+                                                     max_depth = self.hpo_config.model_params['max_depth'],
+                                                     max_features = self.hpo_config.model_params['max_features'],
+                                                     bootstrap = self.hpo_config.model_params['bootstrap'],
+                                                     n_jobs=-1 )\
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/workflows/MLWorkflowSingleCPU.py:110]: Right, both workflows use all the cores during model training, but multi-CPU [i.e., dask] can leverage more parallelization in other parts of the workflow.

File: aws/code/workflows/MLWorkflowMultiCPU.py [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+        """ Close and restart the cluster when multiple cross validation folds are used to prevent memory creep. """    
+        if i_fold == self.hpo_config.cv_folds -1:
+            print('> done all folds; closing cluster\n')
+            await self.client.close()
+            await self.cluster.close()
+        elif i_fold < self.hpo_config.cv_folds - 1:            
+            print('> end of fold; reinitializing cluster\n')
+            await self.client.close()
+            await self.cluster.close()            
+            self.cluster, self.client = self.cluster_initialize()
```
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/workflows/MLWorkflowMultiCPU.py:204]: It was only an issue for multi-GPU RF (at the time 0.14 was the stable branch). There is very minimal overhead for the cluster restart (1 or 2 seconds). I anticipate that we may be able to get rid of the cluster reset in the near future, but its a nice safeguard at the moment.

File: aws/code/train.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/train.py:66]: traceback.print_exc() worked like a charm, thanks!

File: aws/code/train.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (miroenev) [reply aws/code/train.py:37]: Let's come back to this, seems like we need to file a Feature Enhancement for the multi-GPU workflow.

File: aws/code/HPOConfig.py [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+            else:
+                target_files = parquet_files
+            input_file_type = 'Parquet'
+        else:
+            raise Exception("! No [CSV or Parquet] input files detected")
+
+        n_datafiles = len(target_files)
+        assert (n_datafiles > 0)
+
+        pprint.pprint(target_files)
```
Comment (wphicks) [inline aws/code/HPOConfig.py:209]: If you want to move this over to a logging function, [pformat ](https://docs.python.org/3/library/pprint.html#pprint.pformat)will do what you need.

File: aws/helper_functions.py (hunk 7 above)
Comment (wphicks) [reply aws/helper_functions.py:76]: @miroenev Did you miss this one during the update?

**Review by wphicks — CHANGES_REQUESTED** [review]
Two very minor outstanding items. Looking much better overall!

File: aws/code/HPOConfig.py (hunk 26 above)
Comment (wphicks) [reply aws/code/HPOConfig.py:209]: Since we used `pformat` at line 173, do we want to do the same here? Don't mind either way, so long as it's consistent.

File: aws/code/HPOConfig.py [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+            if 'single-CPU' in self.compute_type:
+                model_params.update({'nthreads': os.cpu_count()})
+
+            if 'GPU' in self.compute_type:
+                model_params.update({'tree_method': 'gpu_hist'})
+            else:
+                model_params.update({'tree_method': 'hist'})
+
+        elif 'RandomForest' in self.model_type:
+            # intentionally breaking PEP8 below for argument alignment
```
Comment (wphicks) [inline aws/code/HPOConfig.py:151]: Generally, this is marked with `noqa` directives, since it signals the intention and prevents further linting alarms. No need to change it for this PR; just a note for the future.

**Review by wphicks — CHANGES_REQUESTED** [review]
Looking good! I like all the changes. One final missed comment, and I think we're good to go.

**Review by wphicks — APPROVED** [review]
LGTM!

**Comment by miroenev** [comment]
@JohnZed please merge, thanks!

**Comment by JohnZed** [comment]
Looks great!
